spec/p2p/reactor/p2p-api.md

```diff
@@ -0,0 +1,276 @@
+a method for sending a message to all connected peers:
+
+    func (sw *Switch) Broadcast(e Envelope) chan bool
+
+The `Broadcast()` method is not blocking and returns a channel of booleans.
+For every connected `Peer`, it starts a background thread for sending the
+message to that peer, using the `Peer.Send()` method
+(which is blocking, as detailed in [Send Methods](#send-methods)).
+The result of each unicast send operation (success or failure) is added to the
+returned channel, which is closed when all operations are completed.
+
+> _Note_: while this method is helpful, being adopted by several standard
+> reactors, its current _implementation_ is not efficient, as the marshalling
+> of the provided message is performed as part of the `Peer.Send()` helper
+> method, that is, once per connected peer.
+
+### Vetting Peers
+
+The p2p layer relies on the registered reactors to gauge the _quality_ of peers.
+The following method can be invoked by a reactor to inform the p2p layer that a
+peer has presented a "good" behaviour.
+This information is registered in the node's address book and influences the
+operation of the Peer Exchange (PEX) protocol, as node discovery adopts a bias
+towards "good" peers:
+
+    func (sw *Switch) MarkPeerAsGood(peer Peer)
+
+At the moment, it is up to the consensus reactor to vet a peer.
+A peer is marked as good whenever the consensus protocol collects a multiple of
+`votesToContributeToBecomeGoodPeer = 10000` useful votes or
+`blocksToContributeToBecomeGoodPeer = 10000` useful block parts from that peer.
+A vote or block part are "useful" when they are valid, expected by the node at
+that stage of the protocol, and received for the first time;
+duplicated or redundant messages, fairly common in gossip-based communication,
+are not counted as "useful".
+
+### Stopping Peers
+
+Reactors can instruct the p2p layer to disconnect from a peer.
+Using the p2p layer's nomenclature, the reactor requests a peer to be stopped.
+The peer's send and receive routines are in fact stopped, interrupting the
+communication with the peer.
+The `Peer` is then [removed from every registered reactor][reactor-removepeer],
+using the `RemovePeer(Peer)` method, and from the set of connected peers.
+
+    func (sw *Switch) StopPeerForError(peer Peer, reason interface{})
+
+All the standard reactors employ the above method for disconnecting from a peer
+in case of errors.
+These are errors that occur when processing a message received from a `Peer`.
+The produced `error` is provided to the method as the `reason`.
+
+The `StopPeerForError()` method has an important *caveat*: if the peer to be
+stopped is configured as a _persistent peer_, the switch will attempt
+reconnecting to that same peer.
+While this behaviour makes sense when the method is invoked by other components
+of the p2p layer (e.g., in the case of communication errors), it does not make
+sense when it is invoked by a reactor.
```

Comment:
I guess reconnection might also happen as a result of PEX querying seed nodes for peer addresses, that is, the peer is not put on a "blocked list" for the future, but just removed at this moment. If this is true, perhaps we can add a sentence about that here.